ITISFoundation/osparc-simcore
# ✨ Is677/trial invitations (🔨 app)

## What do these changes do?

We can now create invitation tokens for trial account (i.e. an account remains active for a limited amount of time from its creation).  The trial period is defined in the invitation ``data.trial_account_days`` (see below). If this entry is not defined or set to ``None``, then the account has not expiration date.

###   🔨 devops + application
   - ✨ The invitation link remains the same, e.g.  ``https://example#/registration/?invitation=yVLtenBXr3Cbxo1234aoCr1Wi2d1234``
but the ``confirmation.data``column can define an invitation for a trial account
```json
{
  "created_by": "phillipashley@example.com",
  "guest": "achang@example.org",
  "trial_account_days": 2
}
```
or a permanent account
```json
{
  "created_by": "phillipashley@example.com",
  "guest": "achang@example.org"
}
```
- ✨ @mrnicegyu11, @elisabettai  please note that now we can generate invitation using directly the webserver's app CLI. SEE new command:
```cmd
simcore-service-webserver --help                                        
Usage: simcore-service-webserver [OPTIONS] COMMAND [ARGS]...

Options:
  --install-completion [bash|zsh|fish|powershell|pwsh]
                                  Install completion for the specified shell.
  --show-completion [bash|zsh|fish|powershell|pwsh]
                                  Show completion for the specified shell, to
                                  copy it or customize the installation.
  --help                          Show this message and exit.

Commands:
  invitations  Generates a list of invitation links for registration  <---------------------------
  run            Runs web server
  settings     Resolves settings and prints envfile
```
- 🐛 @odeimaiz  introduced some changes in the Front-end after pair manual testing.

  
## Related issue/s

- follows up from the implementation of PR #3352 
- Closes ITISFoundation/osparc-issues#677

## How to test

Manual test:
- add invitation token in the ``confirmations`` table with an ``trial_account_days=X``
- create invitation link
- use invitation link
  - should create a user
  - ``users.expires_at`` column for that user should set to X days later
- Login as new user
- ``GET v0/me`` should provide the same ``expirationDate``
- Front-end should display expiration date approaches 
- change by hand  ``users.expires_at`` to yesterday
- restart garbage-collector
- ``users.status`` should be automatically set to ``EXPIRED``
- user is unauthorized to do any operation via the API

## Checklist

- [x] check #3352 deployed well in master. Check new column
- [x] manual test with @odeimaiz -> video?
  - ~~logout user in redis~~  Won't do now
  - [x] UI don't show time in preferences @odeimaiz 
  - [x] update invitation script (move to webserver CLI?)
- [x] cleanup code-smells (TODO!)
- [x] check creation of invitations with @mrnicegyu11 @elisabettai 
- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
